File: src/features/typing/data/ts/combination_sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
    const result: number[][] = [];
    const backtrack = (start: number, remaining: number, current: number[]): void => {
        if (remaining === 0) {
            result.push([...current]);
            return;
        }
        for (let i = start; i < candidates.length; i++) {
            if (candidates[i] > remaining) {
                continue;
            }
            current.push(candidates[i]);
            backtrack(i, remaining - candidates[i], current);
            current.pop();
        }
    };
    backtrack(0, target, []);
    return result;
}

File: src/features/typing/data/dart/combination_sum.dart
class Solution {
  List<List<int>> combinationSum(List<int> candidates, int target) {
    List<List<int>> result = [];
    void backtrack(int start, int remaining, List<int> current) {
      if (remaining == 0) {
        result.add(List<int>.from(current));
        return;
      }
      for (var i = start; i < candidates.length; i++) {
        if (candidates[i] > remaining) {
          continue;
        }
        current.add(candidates[i]);
        backtrack(i, remaining - candidates[i], current);
        current.removeLast();
      }
    }
    backtrack(0, target, []);
    return result;
  }
}
